File: src/createInitialPopus.m
% chromos={[FS] [PS] [AS]}
% data=data={1[change_data] 2[job_num] 3[work_num] 4[factory_num] 5[assembly] 6[assembly_data]}
function [chromos] = createInitialPopus(popu, data)
    chromos = cell(popu, 3);
    factory_num = data{1, 4}; [job_num] = data{1, 2}; [work_num] = data{1, 3}; assembly_num = max(data{1, 5});

    for i = 1:popu
        chromos{i, 1} = randi(factory_num, 1, job_num);

        for j = 1:job_num
            chromos{i, 2} = createInitialPopu(work_num, job_num);
        end

        chromos{i, 3} = randperm(assembly_num, assembly_num);
    end

end

% 生成第一个染色体，随机
function initialPopu = createInitialPopu(machNum, workpieceNum)
    lengthChromo = machNum * workpieceNum;
    % 使用 repelem 生成重复的工件编号
    initialPopu = repelem(1:workpieceNum, machNum);
    % 随机排列生成染色体
    initialPopu = initialPopu(randperm(lengthChromo));
end

File: src/createInitialPopus.rs
use rand::seq::SliceRandom;
use rand::Rng;

mod changeDataFunction;


pub fn create_initial_popus(popu: usize, data: &Data) -> Vec<(Vec<usize>, Vec<usize>, Vec<usize>)> {
    let mut chromos = vec![(vec![], vec![], vec![]); popu];
    let factory_num = data.factory_num;
    let job_num = data.job_num;
    let work_num = data.work_num;
    let assembly_num = *data.assembly.iter().max().unwrap() as usize;

    for i in 0..popu {
        chromos[i].0 = (0..job_num).map(|_| rand::thread_rng().gen_range(1..=factory_num)).collect();

        chromos[i].1 = create_initial_popu(work_num, job_num);

        let mut assembly_vec: Vec<usize> = (1..=assembly_num).collect();
        assembly_vec.shuffle(&mut rand::thread_rng());
        chromos[i].2 = assembly_vec;
    }

    chromos
}

fn create_initial_popu(mach_num: usize, workpiece_num: usize) -> Vec<usize> {
    let length_chromo = mach_num * workpiece_num;
    let mut initial_popu: Vec<usize> = (1..=workpiece_num).flat_map(|x| vec![x; mach_num]).collect();
    initial_popu.shuffle(&mut rand::thread_rng());
    initial_popu
}

